File: mcp-server-scaffold/templates/python/template-server-http.py
#!/usr/bin/env python3
"""
{{SERVER_NAME}} MCP Server (HTTP/SSE Transport)

{{DESCRIPTION}}
"""

import asyncio
import json
from contextlib import asynccontextmanager
from collections.abc import AsyncIterator
from mcp.server import Server
from mcp.server.sse import SseServerTransport
from mcp.types import (
    TextContent,
    Tool,
    Resource,
    ErrorCode,
    McpError,
)
from starlette.applications import Starlette
from starlette.routing import Route, Mount
from starlette.requests import Request
from starlette.responses import Response

# Server instance
app = Server("{{SERVER_NAME}}")
sse = SseServerTransport("/messages")

# === TOOL DEFINITIONS ===

TOOLS: list[Tool] = [
    {
        "name": "example_tool",
        "description": "An example tool that demonstrates the pattern",
        "inputSchema": {
            "type": "object",
            "properties": {
                "message": {
                    "type": "string",
                    "description": "Message to echo back",
                },
            },
            "required": ["message"],
        },
    },
]

# === RESOURCE DEFINITIONS ===

RESOURCES: list[Resource] = [
    {
        "uri": "example://docs/readme",
        "name": "Example Resource",
        "mimeType": "text/plain",
        "description": "An example resource demonstrating the pattern",
    },
]

# === HANDLERS ===

@app.list_tools()
async def list_tools() -> list[Tool]:
    """List available tools."""
    return TOOLS

@app.call_tool()
async def call_tool(name: str, arguments: dict) -> list[TextContent]:
    """Execute a tool."""
    try:
        if name == "example_tool":
            message = arguments.get("message", "")
            return [
                TextContent(
                    type="text",
                    text=f"Echo: {message}",
                )
            ]
        else:
            raise McpError(
                ErrorCode.MethodNotFound,
                f"Unknown tool: {name}"
            )
    except McpError:
        raise
    except Exception as e:
        raise McpError(
            ErrorCode.InternalError,
            f"Tool execution failed: {str(e)}"
        )

@app.list_resources()
async def list_resources() -> list[Resource]:
    """List available resources."""
    return RESOURCES

@app.read_resource()
async def read_resource(uri: str) -> str:
    """Read a resource by URI."""
    try:
        if uri == "example://docs/readme":
            return "# Example Resource\n\nThis is a sample resource content."
        else:
            raise McpError(
                ErrorCode.InvalidRequest,
                f"Unknown resource: {uri}"
            )
    except McpError:
        raise
    except Exception as e:
        raise McpError(
            ErrorCode.InternalError,
            f"Resource read failed: {str(e)}"
        )

# === HTTP ROUTES ===

async def handle_sse(request: Request):
    """Handle SSE connection."""
    async with sse.connect_sse(
        request.scope, request.receive, request._send
    ) as streams:
        await app.run(
            streams[0], streams[1],
            app.create_initialization_options()
        )

async def handle_messages(request: Request):
    """Handle POST messages from client."""
    body = await request.body()
    message = json.loads(body)
    
    # Process the message through SSE transport
    response = await sse.handle_message(message)
    
    return Response(
        content=json.dumps(response) if response else "",
        media_type="application/json"
    )

# === STARLETTE APP ===

routes = [
    Route("/sse", endpoint=handle_sse),
    Route("/messages", endpoint=handle_messages, methods=["POST"]),
]

starlette_app = Starlette(routes=routes)

# === SERVER STARTUP ===

if __name__ == "__main__":
    import uvicorn
    
    print("{{SERVER_NAME}} MCP server starting on http://localhost:3000")
    print("SSE endpoint: GET /sse")
    print("Message endpoint: POST /messages")
    
    uvicorn.run(starlette_app, host="0.0.0.0", port=3000)

File: skills/mcp-server-scaffold/templates/typescript/template-server-http.ts
import express, { Request, Response } from "express";
import { SSEServerTransport } from "@modelcontextprotocol/sdk/server/sse.js";
import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import {
  CallToolRequestSchema,
  ListToolsRequestSchema,
  ListResourcesRequestSchema,
  ReadResourceRequestSchema,
  ErrorCode,
  McpError,
} from "@modelcontextprotocol/sdk/types.js";

// Import tool handlers
// import { handleGetForecast } from "./tools/get-forecast.js";

// Import resource providers  
// import { handleQueryResource } from "./resources/query.js";

/**
 * {{SERVER_NAME}} MCP Server (HTTP/SSE Transport)
 * 
 * {{DESCRIPTION}}
 * 
 * Environment:
 *   PORT - Server port (default: 3000)
 *   CORS_ORIGIN - Allowed CORS origin (default: *)
 */

const app = express();
const PORT = process.env.PORT || 3000;
const CORS_ORIGIN = process.env.CORS_ORIGIN || "*";

// Enable CORS for browser-based clients
app.use((req, res, next) => {
  res.header("Access-Control-Allow-Origin", CORS_ORIGIN);
  res.header("Access-Control-Allow-Headers", "*");
  res.header("Access-Control-Allow-Methods", "GET, POST, OPTIONS");
  if (req.method === "OPTIONS") {
    res.sendStatus(200);
  } else {
    next();
  }
});

app.use(express.json());

// Map to store transports by session ID
const transports: Map<string, SSEServerTransport> = new Map();

/**
 * Create MCP server instance
 */
function createServer(): Server {
  const server = new Server(
    {
      name: "{{SERVER_NAME}}",
      version: "{{VERSION}}",
    },
    {
      capabilities: {
        tools: {},
        resources: {},
      },
    }
  );

  // === TOOL HANDLERS ===

  server.setRequestHandler(ListToolsRequestSchema, async () => {
    return {
      tools: [
        {
          name: "example_tool",
          description: "An example tool that demonstrates the pattern",
          inputSchema: {
            type: "object",
            properties: {
              message: {
                type: "string",
                description: "Message to echo back",
              },
            },
            required: ["message"],
          },
        },
      ],
    };
  });

  server.setRequestHandler(CallToolRequestSchema, async (request) => {
    const { name, arguments: args } = request.params;

    try {
      switch (name) {
        case "example_tool": {
          const message = args?.message as string;
          return {
            content: [
              {
                type: "text",
                text: `Echo: ${message}`,
              },
            ],
          };
        }

        default:
          throw new McpError(
            ErrorCode.MethodNotFound,
            `Unknown tool: ${name}`
          );
      }
    } catch (error) {
      if (error instanceof McpError) {
        throw error;
      }
      throw new McpError(
        ErrorCode.InternalError,
        `Tool execution failed: ${error}`
      );
    }
  });

  // === RESOURCE HANDLERS ===

  server.setRequestHandler(ListResourcesRequestSchema, async () => {
    return {
      resources: [
        {
          uri: "example://docs/readme",
          name: "Example Resource",
          mimeType: "text/plain",
          description: "An example resource demonstrating the pattern",
        },
      ],
    };
  });

  server.setRequestHandler(ReadResourceRequestSchema, async (request) => {
    const uri = request.params.uri;

    try {
      switch (uri) {
        case "example://docs/readme": {
          return {
            contents: [
              {
                uri,
                mimeType: "text/plain",
                text: "# Example Resource\n\nThis is a sample resource content.",
              },
            ],
          };
        }

        default:
          throw new McpError(
            ErrorCode.InvalidRequest,
            `Unknown resource: ${uri}`
          );
      }
    } catch (error) {
      if (error instanceof McpError) {
        throw error;
      }
      throw new McpError(
        ErrorCode.InternalError,
        `Resource read failed: ${error}`
      );
    }
  });

  return server;
}

// === SSE ENDPOINTS ===

// SSE connection endpoint
app.get("/sse", async (req: Request, res: Response) => {
  const sessionId = req.query.sessionId as string || crypto.randomUUID();
  
  const transport = new SSEServerTransport("/message", res);
  transports.set(sessionId, transport);
  
  const server = createServer();
  
  // Clean up on disconnect
  transport.onclose = () => {
    transports.delete(sessionId);
    console.error(`Session ${sessionId} closed`);
  };
  
  await server.connect(transport);
  console.error(`Session ${sessionId} connected`);
});

// Message endpoint (POST from client)
app.post("/message", async (req: Request, res: Response) => {
  const sessionId = req.query.sessionId as string;
  
  if (!sessionId) {
    res.status(400).json({ error: "Missing sessionId" });
    return;
  }
  
  const transport = transports.get(sessionId);
  if (!transport) {
    res.status(404).json({ error: "Session not found" });
    return;
  }
  
  await transport.handlePostMessage(req, res);
});

// Health check endpoint
app.get("/health", (_req: Request, res: Response) => {
  res.json({ 
    status: "ok", 
    server: "{{SERVER_NAME}}",
    version: "{{VERSION}}",
    activeSessions: transports.size 
  });
});

// Start server
app.listen(PORT, () => {
  console.error(`{{SERVER_NAME}} MCP server running on http://localhost:${PORT}`);
  console.error(`SSE endpoint: http://localhost:${PORT}/sse`);
  console.error(`Health check: http://localhost:${PORT}/health`);
});
